presto-main/src/main/java/io/prestosql/sql/planner/OrExpressionNormalizer.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Multimap;
+import io.prestosql.metadata.Metadata;
+import io.prestosql.sql.tree.ComparisonExpression;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.ExpressionRewriter;
+import io.prestosql.sql.tree.ExpressionTreeRewriter;
+import io.prestosql.sql.tree.InListExpression;
+import io.prestosql.sql.tree.InPredicate;
+import io.prestosql.sql.tree.LogicalBinaryExpression;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.prestosql.sql.ExpressionUtils.combineDisjuncts;
+import static io.prestosql.sql.ExpressionUtils.or;
+import static io.prestosql.sql.tree.BooleanLiteral.FALSE_LITERAL;
+import static io.prestosql.sql.tree.ComparisonExpression.Operator.EQUAL;
+import static io.prestosql.sql.tree.LogicalBinaryExpression.Operator.AND;
+import static io.prestosql.sql.tree.LogicalBinaryExpression.Operator.OR;
+
+public class OrExpressionNormalizer
+{
+    private OrExpressionNormalizer()
+    {}
+
+    public static Expression normalizeOrExpressions(Metadata metadata, Expression expression)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(metadata), expression, Optional.empty());
```

Comment:
To me it is a bit misleading that rewrite sometimes does not rewrite but only collect information in context. Maybe you could do it in two phases.  For every logical binary expression you could collect with simple visitor (not rewriter) expressions. So rewriting visitor would implement only `rewriteLogicalBinaryExpression`.

Reply:
https://github.com/prestosql/presto/pull/2698#issuecomment-581424354